explorer/interpreter/type_checker.cpp

```diff
@@ -987,8 +1003,13 @@ auto TypeChecker::ArgumentDeduction::Deduce(Nonnull<const Value*> param,
       if (arg->kind() != Value::Kind::PointerType) {
         return handle_non_deduced_type();
       }
-      return Deduce(&cast<PointerType>(*param).type(),
-                    &cast<PointerType>(*arg).type(),
+      const auto& param_pointed = cast<PointerType>(param)->type();
+      const auto& arg_pointed = cast<PointerType>(arg)->type();
+      if (const auto* arg_class = dyn_cast<NominalClassType>(&arg_pointed);
+          arg_class && arg_class->InheritsClass(&param_pointed)) {
```

Comment:
We'll also need to decide if we want a C++-like deduce-from-base-class rule:
```
base class B(T:! Type) {}
class D extends B(i32) {}
fn F[T:! Type](p: B(T)*) {}
fn Main() {
  var d: D = {};
  // Is this OK, deducing T = i32, or an error?
  F(&d);
}
```
I've filed this as #2464. I'm happy for us to disallow this for now (as this change does), but I think it'd be useful to add a TODO mentioning that issue and the above testcase or something like it.

Reply:
Added